File: py/creational/factory_method.py
# Factory Method lets a superclass workflow defer object creation to subclasses.
# Modern Python note: passing a constructor or callable is often simpler than a
# subclass hierarchy unless the whole workflow is meant to be specialized.
from abc import ABC, abstractmethod


class Document(ABC):
    @abstractmethod
    def open(self):
        pass


class PdfDocument(Document):
    def open(self):
        return "opening PDF"


class TextDocument(Document):
    def open(self):
        return "opening text file"


class Application(ABC):
    def open_document(self):
        document = self.create_document()
        return document.open()

    @abstractmethod
    def create_document(self):
        pass


class PdfApplication(Application):
    def create_document(self):
        return PdfDocument()


class TextApplication(Application):
    def create_document(self):
        return TextDocument()


if __name__ == "__main__":
    print(PdfApplication().open_document())

File: csharp/creational/factory_method.cs
namespace Gof.ModernCSharp.Creational;

// Factory Method lets a superclass workflow defer object creation to subclasses.
// Modern C# note: passing Func<T> or using DI is often simpler unless the whole
// workflow is intentionally specialized by inheritance.
public static class FactoryMethod
{
    interface IDocument { string Open(); }
    sealed class PdfDocument : IDocument { public string Open() => "opening PDF"; }

    abstract class Application
    {
        public string OpenDocument() => CreateDocument().Open();
        protected abstract IDocument CreateDocument();
    }

    sealed class PdfApplication : Application
    {
        protected override IDocument CreateDocument() => new PdfDocument();
    }

    public static string Run() => new PdfApplication().OpenDocument();
}
